R/build_browser_hub.R

```diff
@@ -0,0 +1,286 @@
+                col.names = FALSE)
+    if(projection == "grch37"){
+      chr_arms = GAMBLR.data::chromosome_arms_grch37 %>% 
+        mutate(chromosome = paste0("chr", chromosome))
+    }else{ # so projection is hg38
+      chr_arms = GAMBLR.data::chromosome_arms_hg38
+    }
+    chr_sizes = chr_arms %>%
+      dplyr::filter(arm == "q") %>%
+      dplyr::select(chromosome, end) %>%
+      rename(size = "end")
+    temp_chr_sizes = tempfile(pattern = "chrom.sizes_")
+    write.table(chr_sizes, temp_chr_sizes, quote = FALSE, sep = "\t", row.names = FALSE, 
+                col.names = FALSE)
+    bedtobigbed = GAMBLR.helpers::check_config_value(config::get("dependencies")$bedToBigBed)
+    regions_bed_file = file.path(track_dir, "regions.bb")
+    bigbed_conversion = gettextf("%s %s %s %s", bedtobigbed, temp_bed, temp_chr_sizes, 
+                                 regions_bed_file)
+    system(bigbed_conversion)
+    unlink(temp_bed)
+    unlink(temp_chr_sizes)
+  }else{
+    regions_bed_file = file.path(track_dir, "regions.bed")
+    write.table(regions_bed, regions_bed_file, quote = FALSE, sep = "\t", row.names = FALSE, 
+                col.names = FALSE)
+  }
+  
+  # get maf data from the specified regions and metadata/samples (for each seq type)
+  maf_data = mapply(function(these_seq_types_i, these_samples_metadata_i){
+    get_ssm_by_regions(regions_bed = regions_bed, this_seq_type = these_seq_types_i,
+                       these_samples_metadata = these_samples_metadata_i, 
+                       projection = projection, streamlined = FALSE, 
+                       basic_columns = TRUE) %>% 
+      suppressMessages
+  }, these_seq_types, these_samples_metadata, SIMPLIFY = FALSE)
+  
+  # split maf table according to splitColumnName
+  if(!is.null(splitColumnName)){
+    maf_data = mapply(function(maf_data_i, these_samples_metadata_i){
+      maf_data_i = dplyr::select(these_samples_metadata_i, sample_id, all_of(splitColumnName)) %>% 
+        distinct(sample_id, .keep_all = TRUE) %>% 
+        left_join(maf_data_i, ., join_by(Tumor_Sample_Barcode == sample_id))
+      dplyr::select(maf_data_i, -all_of(splitColumnName)) %>% 
+        split(maf_data_i[[splitColumnName]])
+    }, maf_data, these_samples_metadata)
+    track_names = lapply(maf_data, names)
+    these_samples_metadata = mapply(function(these_samples_metadata_i, track_names_i){
+      split(these_samples_metadata_i, these_samples_metadata_i[[splitColumnName]]) %>% 
+        "["(track_names_i)
+    }, these_samples_metadata, track_names)
+  }else{
+    maf_data = lapply(maf_data, list)
+    track_names = lapply(these_seq_types, function(x) list("all"))
+    these_samples_metadata = lapply(these_samples_metadata, list)
+  }
+  
+  # check seq types that ssms could be found
+  is_there_muts = lengths(track_names) > 0
+  if( !any(is_there_muts) ){
+    stop("No SSMs were found. Reset your paramters.")
+  }
+  maf_data = maf_data[is_there_muts]
```

Comment:
Sorry, I think I am bit lost and just trying to confirm it with you. The `is_there_muts` variable is logical here, right? So how is it used in the subletting of the maf and other data below this line?

Reply:
Yes, `is_there_muts` is logical and it checks whether there are mutations from each provided seq_type. 
```
> is_there_muts = lengths(track_names) > 0
> is_there_muts
 genome capture 
   TRUE   FALSE 
```
In this example, since no mutation was found from capture, we need to remove this seq type. 
```
> class(maf_data)
[1] "list"
> names(maf_data)
[1] "genome"  "capture"
> maf_data$capture
named list()

maf_data = maf_data[is_there_muts]   # remove capture because it's empty
```
The same is done for the other objects besides `maf_data`. All these objects are lists or vectors (not data frames).